Add renderer option to storyshots

tobilen opened the pull request:
Issue: #2408

## What I did
Added two options to the options passed to `initStoryshots`, `renderer` and `serializer`. Their usage is documented in the readme, but in short, you can pass custom renderer and serializer functions to the storyshot addon.
This is probably a bit easier to use than the `test` option, which requires you to pass an entire snapshotting test.

## How to test
In addition to the storyshot tests currently in the addon, i added one where enzymes `mount` renderer and the serializer from `enzyme-to-json` are passed.

Does this need a new example in the kitchen sink apps?
No

Does this need an update to the documentation?
Yes, which has been added.

tobilen committed:
Add option in snapshot generator to use renderer from options if passed

tobilen committed:
Use snapshot with options as default, pass on renderer to test-bodies

tobilen committed:
Add option to pass serializer to storyshots

tobilen committed:
Add enzyme-to-json for serializing, initStoryshots with custom renderer

tobilen committed:
Extend readme

tobilen committed:
Move enzyme-to-json dependency to storyshots, split initStoryshots tests

tobilen committed:
Remove react-test-renderer storyshot snapshots, since they dont work yet

Hypnosphi commented on addons/storyshots/.babelrc:
```diff
@@ -1,5 +1,5 @@
 {
-  "presets": ["env", "react"],
+  "presets": ["env", "stage-0", "react"],
```
Why is it needed? Can we include the particular transform instead?

tobilen replied:
for the destructuring assignment here: https://github.com/storybooks/storybook/pull/2414/files#diff-75fdf0f2c6757366489b97c6898ea585R6

i would have used stage-2, but storybook itself pulls in [stage-0](https://github.com/storybooks/storybook/blob/master/.babelrc#L2), so i went with that to stay consistent

Hypnosphi replied:
Oh I see. I wonder if we can remove this nested `.babelrc` to avoid confusion 

@ndelangen what do you think?

Hypnosphi commented:
Looks great! Let's use this feature for storyshots of `example/cra-kitchen-sink`

tobilen committed:
Add dependencies necessary to use enzyme in jest

tobilen committed:
Use storyshots renderer option to use enzyme

tobilen committed:
Use storyshots renderer option to use enzyme, add snapshots to project

tobilen reviewed (commented):
i've extended the demo project to use it. i went for jest configuration in package.json over initializing everything in the initSnapshot file, since that is the cleaner way to do it imo

Hypnosphi commented on examples/cra-kitchen-sink/src/enzyme.js:
```diff
@@ -0,0 +1,4 @@
+import Enzyme from 'enzyme';
+import Adapter from 'enzyme-adapter-react-16';
+
+Enzyme.configure({ adapter: new Adapter() });
```
We have this setup on root level, no need to duplicate it here

tobilen replied:
same as https://github.com/storybooks/storybook/pull/2414#discussion_r154497265

Hypnosphi commented on examples/cra-kitchen-sink/package.json:
```diff
@@ -39,6 +39,9 @@
     "@storybook/components": "^3.3.0-alpha.4",
     "@storybook/react": "^3.3.0-alpha.4",
     "babel-jest": "^21.2.0",
+    "enzyme": "^3.2.0",
```
We need to have a single enzyme instance across the codebase, so it shouldn’t be added to package dependencies

tobilen replied:
i know. but i figured an "example project" should include the necessary dependencies? its a lot more difficult to understand if i need to look outside of the project for its code to work. thats why i included enzyme setup there

Hypnosphi replied:
OK, makes sense

Hypnosphi commented on examples/cra-kitchen-sink/src/storyshots.test.js:
```diff
@@ -1,17 +1,9 @@
-  test: multiSnapshotWithOptions({
-    createNodeMock,
-  }),
+  renderer,
```
Can we support custom renderers for multisnapshots as well?

tobilen replied:
yeah sure. done

Hypnosphi replied:
Looks like it still stores all the snapshots in one file

tobilen replied:
it does. i'm currently debugging it. looks like it did not work in the first place.  the multisnapshot function doesnt seem to be called correctly on the `release/3.3` branch

tobilen replied:
should work now. we cant use enzymes `mount` since that renders out the context prop (includes the file path, which is different on every machine). We also need to pass a serializer since jest-specific-snapshots doesnt use the projects serializer

tobilen committed:
Use multisnapshots in test project

tobilen committed:
Remove stored snapshots, since multi snapshots are different on the CI

tobilen committed:
Use enzymes render renderer.

mount includes context props, which means current file paths will be printed to snapshots. those are machine-dependant and differ on every enviroment. we dont want that.

tobilen committed:
Correctly use multiSnapshotTest in example project

Hypnosphi commented on examples/cra-kitchen-sink/src/stories/__snapshots__/addon-a11y.stories.storyshot:
```diff
@@ -1,46 +1,29 @@
-  onClick={[Function]}
-  style={Object {}}
->
+<button>
```
Do you have an idea why Enzyme strips attributes?

tobilen replied:
its the renderer specified in the tests. you have `shallow` which shallowrenders, `mount` which more or less renders the entire vdom, and `render` (uses cheerio under the  hood) which just prints out html. see https://github.com/airbnb/enzyme/blob/master/docs/api/render.md

like i [said](https://github.com/storybooks/storybook/pull/2414#discussion_r154498667), we cant use `mount` with multisnapshot because the snapshots then include the `context` prop, and part of that is the file path. and the file path changes depending on what the absolute path on the system the snapshot is stored at

tobilen replied:
im not very familiar with the internals of jest-specific-snapshots, but if there was a way for it to pass a relative path in the context, that would eliminate the problem. its where i see the solution to the problem at least.
i wouldnt want to modify the output of the renderer (like replacing the filepath with an empty string or somesuch), which is the only alternative i can think of

tobilen committed:
Merge branch 'release/3.3' into add-renderer-option-to-storyshots
